Assets/Features/Telemetry Lap File/TelemetryLapFileWriter.cs

```diff
@@ -0,0 +1,186 @@
+    {
+        private CSVFileWriter fileWriter;
+        public string Filename { get; private set; }
+        public string FullRelativePath { get; private set; }
+        public string FullPath { get; private set; }
+        public string MetadataFullRelativePath { get; private set; }
+        public string TempFullRelativePath { get; private set; }
+        public bool HeadersWritten { get; private set; }
+        public int ColumnCount { get; private set; }
+        public int LineCount { get; private set; }
+
+        private const string root = "Telemetry";
+        private const string separator = ",";
+        private readonly IFormatProvider invariantCulture;
+        private readonly StringBuilder builder;
+        private readonly TimeFormatter timeFormatter;
+
+
+        public IReadOnlyList<string> Headers { get; private set; }
+        public IReadOnlyList<string> Units { get; private set; }
+        public bool IsRecordingReady { get; private set; }
+
+        public TelemetryLapFileWriter(IReadOnlyList<string> headers, IReadOnlyList<string> units)
+        {
+            this.Headers = headers;
+            this.Units = units;
+
+            invariantCulture = System.Globalization.CultureInfo.InvariantCulture;
+            builder = new StringBuilder();
+            timeFormatter = new TimeFormatter(TimeFormatter.Mode.MinutesAndSeconds, @"mm\.ss\.fff", @"ss\.fff");
+
+            if (!Directory.Exists(root))
+            {
+                Directory.CreateDirectory(root);
+            }
+        }
+
+        public void StartRecording()
+        {
+            TempFullRelativePath = Path.Combine(root, "temp.csv");
+
+            FileStream fs = new FileStream(TempFullRelativePath, FileMode.Create, FileAccess.Write);
+            fileWriter = new CSVFileWriter(fs);
+            
+            HeadersWritten = false;
+            WriteHeaders(Headers, Units);
+            
+            IsRecordingReady = true;
+        }
+
+        public void StopRecordingAndSaveFile(TelemetryLapMetadata meta)
+        {
+            SetFileNames(meta);
+            DisposeFileAndRename();
+            Writemetadata(meta);
+
+            IsRecordingReady = false;
+        }
+
+        private void SetFileNames(TelemetryLapMetadata meta)
+        {
+            string dateStr = DateTime.UtcNow.ToString("yyyy-MM-dd HH.mm.ss UTC", invariantCulture);
+            string lapTimeStr = meta.completed ? timeFormatter.ToString(meta.lapTime) : "unfinished";
+            string synthetic = meta.ideal ? " ideal" : string.Empty;
+
+            Filename = $"{dateStr} {lapTimeStr}{synthetic}.csv";
+            FullRelativePath = Path.Combine(root, Filename);
+            FullPath = Path.Combine(Application.dataPath, FullRelativePath);
+        }
+        private void DisposeFileAndRename()
+        {
+            fileWriter.Dispose();
+            File.Move(TempFullRelativePath, FullRelativePath); // Rename the oldFileName into newFileName
```

Comment:
Why rename? Why cannot keep the old filename?

Reply:
The name of the file is {timestamp} {laptime}.csv. When we start writing the file, we don't  know neither of the two. So we start writing the file with the name "temp.csv". When we complete the lap, we do know both, the timestamp and the laptime, so we rename the "temp.csv" file by its final name

Reply:
awesome